### PR TITLE
Added advanced user model with CURD written in promise style

### PR DESCRIPTION
**Pull request by RonaldChao**

### CODE REVIEW

File: routers/advancedusers.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
+// Reference
+// https://github.com/globalbanana/website/blob/dev/src/router/login.js
+
+const express = require("express");
+const router = express.Router();
+
+const advancedUserModel = require("../modules/dataBase/advancedUsers.js");
+
+// Retrieve the user record
+router.get("/get/:username", (req, res) => {
```
Comment (wahengchang) [inline routers/advancedusers.js]: just use `router.get(/:id)`, to keep RESTful style

File: routers/advancedusers.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  advancedUserModel
+    .getByCondition(condition)
+    .then(obj => res.send(`Your name is ${obj.name}, ok?`))
+    .catch(err =>
+      res.send(`The username ${req.params.username} does not exists`)
+    );
+});
+
+// Register a new user
+router.post("/register", (req, res) => {
```
Comment (wahengchang) [inline routers/advancedusers.js]: just use router.post(/), to keep RESTful style

File: routers/advancedusers.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  advancedUserModel
+    .create(data)
+    .then(
+      () => res.json({ success: true, msg: "Registered user" }),
+      err => res.json({ success: false, msg: "Failed to register" })
+    );
+});
+
+// Edit the user profile
+router.put("/edit", (req, res) => {
```
Comment (wahengchang) [inline routers/advancedusers.js]: just use router.put(/:id), to keep RESTful style

File: routers/advancedusers.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+  advancedUserModel
+    .updateByCondition(condition, update, null)
+    .then(
+      () => res.json({ success: true, msg: "Updated user" }),
+      err => res.json({ success: false, msg: "Failed to update user" })
+    );
+});
+
+// Delete a user
```
Comment (wahengchang) [inline routers/advancedusers.js]: just use router.delete(/:id), to keep RESTful style

File: routers/users.js [hunk 5]
```diff
@@ -1,16 +1,69 @@
...
-const router = express.Router()
+const mongoose = require("mongoose");
+const User = mongoose.model("Users");
 
-router.get('/', (req, res) => {  
-    const fakeUser = {
-        id: 'fakeid',
-        username: 'peterchang'
+// Retrieve the user record
+router.get("/get/:username", (req, res) => {
```
Comment (wahengchang) [inline routers/users.js]: just use router.get(/:id), to keep RESTful style

**Comment by wahengchang** [comment]
what is different between `user` and `advanceUser`?

**Comment by RonaldChao** [comment]
Not many meaningful difference.
The difference lies in the way of performing CURD
User do everything (e.g. find(), findAndUpdate()) within the router's route handler. 
The CURD operation for advancedUser is written in Promise. An abstract class is also created for generic database communication.

**Comment by wahengchang** [comment]
HTTP 原本 support 的動作

```
GET /api/v1/users (list)
GET /api/v1/users/id
PUT /api/v1/users/id
POST /api/v1/users/id
DELETE /api/v1/users/id
```

File: app.js [hunk 6]
```diff
@@ -1,7 +1,24 @@
-var express = require('express')
-const port = process.env.PORT || 3000
-
-var app = express()
+var express = require("express");
+const port = process.env.PORT || 3000;
+const bodyParser = require("body-parser");
+const mongoose = require("mongoose");
```
Comment (wahengchang) [inline app.js]: remove no-use code

File: app.js [hunk 7]
```diff
@@ -1,7 +1,24 @@
...
+const mongoUsername = process.env.MONGO_USERNAME;
+const mongoPassword = process.env.MONGO_PASSWORD;
+const mongoURL = process.env.MONGO_URL;
+const databaseKey = `mongodb://${mongoUsername}:${mongoPassword}@${mongoURL}`;
+
+require("./modules/dataBase/user");
```
Comment (wahengchang) [inline app.js]: use `'` instend of `"`

File: app.js [hunk 8]
```diff
@@ -22,20 +39,23 @@ var app = express()
...
 //  Knowledge
 //      MongoLab
 //      bike (CURD: Create, Update, Read, Delete)
 
-app.use('/users', require('./routers/users'))
+// Install middleware to parse the request body
+app.use(bodyParser.json());
+
+// app.use("/users", require("./routers/users"));
+app.use("/api/v1/advancedusers", require("./routers/advancedusers"));
```
Comment (wahengchang) [inline app.js]: use ' instend of "

File: modules/dataBase/abstract.js [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+  const mongoUsername = process.env.MONGO_USERNAME;
+  const mongoPassword = process.env.MONGO_PASSWORD;
+  const mongoURL = process.env.MONGO_URL;
+  const databaseKey = `mongodb://${mongoUsername}:${mongoPassword}@${mongoURL}`;
+
+  mongoose.connect(databaseKey, err => {
+    if (err) {
+      console.log(err);
+    } else {
+      console.log("Connected (Abstract)");
```
Comment (wahengchang) [inline modules/dataBase/abstract.js:14]: use ' instend of "

File: modules/dataBase/advancedUsers.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    unqiue: true
+  },
+  username: {
+    type: String,
+    required: true,
+    unique: true
+  },
+  gender: {
+    type: String,
+    default: "N/A"
```
Comment (wahengchang) [inline modules/dataBase/advancedUsers.js]: is `"N/A"` a default string ?

**Review by wahengchang — APPROVED** [review]
👍

**Pull request merged**